在 Ubuntu jammy 上 pytest 失败

## Ruyi 版本

+ 0.37.1

## 问题描述

在 Ubuntu Jammy 上 pytest 失败

```
$ python -m pytest -v
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.10.12, pytest-6.2.5, py-1.10.0, pluggy-0.13.0 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /home/hachi/ruyi-pkg/ruyi/ruyi-0.37.1, configfile: pyproject.toml, testpaths: tests
plugins: mock-3.6.1
collected 39 items                                                                                                                                                                                                                         

tests/config/test_editor.py::test_enter_exit PASSED                                                                                                                                                                                  [  2%]
tests/config/test_editor.py::test_set_value PASSED                                                                                                                                                                                   [  5%]
tests/config/test_editor.py::test_unset_value_remove_section PASSED                                                                                                                                                                  [  7%]
tests/config/test_editor.py::test_malformed_config_file_error PASSED                                                                                                                                                                 [ 10%]
tests/config/test_schema.py::test_decode_value_bool PASSED                                                                                                                                                                           [ 12%]
tests/config/test_schema.py::test_decode_value_str PASSED                                                                                                                                                                            [ 15%]
tests/config/test_schema.py::test_decode_value_datetime PASSED                                                                                                                                                                       [ 17%]
tests/config/test_schema.py::test_encode_value_bool PASSED                                                                                                                                                                           [ 20%]
tests/config/test_schema.py::test_encode_value_int PASSED                                                                                                                                                                            [ 23%]
tests/config/test_schema.py::test_encode_value_str PASSED                                                                                                                                                                            [ 25%]
tests/config/test_schema.py::test_encode_value_datetime PASSED                                                                                                                                                                       [ 28%]
tests/pluginhost/test_api.py::test_api_with_ FAILED                                                                                                                                                                                  [ 30%]
tests/ruyipkg/test_checksum.py::test_get_hash_instance_supported PASSED                                                                                                                                                              [ 33%]
tests/ruyipkg/test_checksum.py::test_get_hash_instance_unsupported PASSED                                                                                                                                                            [ 35%]
tests/ruyipkg/test_checksum.py::test_checksummer_compute PASSED                                                                                                                                                                      [ 38%]
tests/ruyipkg/test_checksum.py::test_checksummer_check PASSED                                                                                                                                                                        [ 41%]
tests/ruyipkg/test_entity.py::test_entity_store_discovery FAILED                                                                                                                                                                     [ 43%]
tests/ruyipkg/test_entity.py::test_entity_store_get_entity FAILED                                                                                                                                                                    [ 46%]
tests/ruyipkg/test_entity.py::test_entity_store_iter_entities FAILED                                                                                                                                                                 [ 48%]
tests/ruyipkg/test_entity.py::test_entity_store_get_entity_by_ref FAILED                                                                                                                                                             [ 51%]
tests/ruyipkg/test_entity.py::test_entity_validation FAILED                                                                                                                                                                          [ 53%]
tests/ruyipkg/test_entity.py::test_entity_related_refs FAILED                                                                                                                                                                        [ 56%]
tests/ruyipkg/test_entity.py::test_get_related_entities FAILED                                                                                                                                                                       [ 58%]
tests/ruyipkg/test_entity.py::test_traverse_related_entities_direct FAILED                                                                                                                                                           [ 61%]
tests/ruyipkg/test_entity.py::test_traverse_related_entities_transitive FAILED                                                                                                                                                       [ 64%]
tests/ruyipkg/test_entity.py::test_traverse_related_entities_with_type_filter FAILED                                                                                                                                                 [ 66%]
tests/ruyipkg/test_entity.py::test_entity_store_is_entity_related_to FAILED                                                                                                                                                          [ 69%]
tests/ruyipkg/test_entity.py::test_entity_porcelain_output FAILED                                                                                                                                                                    [ 71%]
tests/ruyipkg/test_entity_providers.py::test_entity_store_with_custom_provider FAILED                                                                                                                                                [ 74%]
tests/ruyipkg/test_format_manifest.py::test_format_manifest FAILED                                                                                                                                                                   [ 76%]
tests/ruyipkg/test_host.py::test_canonicalize_arch PASSED                                                                                                                                                                            [ 79%]
tests/ruyipkg/test_host.py::test_canonicalize_host PASSED                                                                                                                                                                            [ 82%]
tests/ruyipkg/test_state.py::test_bound_installation_state_store_empty PASSED                                                                                                                                                        [ 84%]
tests/ruyipkg/test_state.py::test_bound_installation_state_store_with_installed_packages PASSED                                                                                                                                      [ 87%]
tests/ruyipkg/test_unpack.py::test_unpack_deb FAILED                                                                                                                                                                                 [ 89%]
tests/utils/test_l10n.py::test_lang_code_to_lang_region PASSED                                                                                                                                                                       [ 92%]
tests/utils/test_l10n.py::test_match_lang_code PASSED                                                                                                                                                                                [ 94%]
tests/utils/test_toml.py::test_extract_header_comments PASSED                                                                                                                                                                        [ 97%]
tests/utils/test_toml.py::test_extract_footer_comments PASSED                                                                                                                                                                        [100%]

================================================================================================================= FAILURES =================================================================================================================
______________________________________________________________________________________________________________ test_api_with_ ______________________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ab4b130>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ab4b160>

    def test_api_with_(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        class MockContextManager(AbstractContextManager[int]):
            def __init__(self) -> None:
                self.entered = 0
                self.exited = 0
    
            def __enter__(self) -> int:
                self.entered += 1
                return 233
    
            def __exit__(
                self,
                exc_type: type[BaseException] | None,
                exc_value: BaseException | None,
                traceback: TracebackType | None,
            ) -> bool | None:
                self.exited += 1
                return None
    
>       with ruyi_file.plugin_suite("with_") as plugin_root:

tests/pluginhost/test_api.py:34: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.10/contextlib.py:135: in __enter__
    return next(self.gen)
/usr/lib/python3.10/importlib/_common.py:92: in _tempfile
    os.write(fd, reader())
/usr/lib/python3.10/importlib/abc.py:371: in read_bytes
    with self.open('rb') as strm:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <importlib._adapters.DegenerateFiles.Path object at 0x7f9b4ab8c280>, mode = 'rb', args = (), kwargs = {}

    def open(self, mode='rb', *args, **kwargs):
>       raise ValueError()
E       ValueError

/usr/lib/python3.10/importlib/_adapters.py:54: ValueError
_______________________________________________________________________________________________________ test_entity_store_discovery ________________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ba063e0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ba05db0>

    def test_entity_store_discovery(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test that EntityStore correctly discovers entity types."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ba063e0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
_______________________________________________________________________________________________________ test_entity_store_get_entity _______________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4c7d53c0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4c7d5de0>

    def test_entity_store_get_entity(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test retrieving entities by type and ID."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4c7d53c0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
_____________________________________________________________________________________________________ test_entity_store_iter_entities ______________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ab4ac20>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ab4ab30>

    def test_entity_store_iter_entities(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test iterating over entities."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ab4ac20>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
___________________________________________________________________________________________________ test_entity_store_get_entity_by_ref ____________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4c7d54e0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ada91b0>

    def test_entity_store_get_entity_by_ref(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test retrieving entities by reference string."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:68: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4c7d54e0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
__________________________________________________________________________________________________________ test_entity_validation __________________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4aa10a90>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4aa12b90>

    def test_entity_validation(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test entity validation against schemas."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4aa10a90>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
_________________________________________________________________________________________________________ test_entity_related_refs _________________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4aa53df0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4aa51060>

    def test_entity_related_refs(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test retrieving related entity references from an entity."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:106: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4aa53df0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
________________________________________________________________________________________________________ test_get_related_entities _________________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4c7d53c0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4c7d54e0>

    def test_get_related_entities(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test retrieving related entities from an entity."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:127: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4c7d53c0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
__________________________________________________________________________________________________ test_traverse_related_entities_direct ___________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ab8f1c0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ab8fbb0>

    def test_traverse_related_entities_direct(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test traversing directly related entities."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:154: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ab8f1c0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
________________________________________________________________________________________________ test_traverse_related_entities_transitive _________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4a9dadd0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4a9d8dc0>

    def test_traverse_related_entities_transitive(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test traversing the transitive closure of related entities."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:175: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4a9dadd0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
_____________________________________________________________________________________________ test_traverse_related_entities_with_type_filter ______________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ada9c00>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ada99f0>

    def test_traverse_related_entities_with_type_filter(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test traversing related entities with filtering by entity type."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:211: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ada9c00>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
__________________________________________________________________________________________________ test_entity_store_is_entity_related_to __________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ba05e70>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ba06500>

    def test_entity_store_is_entity_related_to(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test the ``is_related_to`` method of ``EntityStore``."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:261: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ba05e70>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
_______________________________________________________________________________________________________ test_entity_porcelain_output _______________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ab8dde0>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4ab8cbb0>

    def test_entity_porcelain_output(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test that entity porcelain output works correctly."""
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity.py:329: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4ab8dde0>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
__________________________________________________________________________________________________ test_entity_store_with_custom_provider __________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4aa12380>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4aa13670>

    def test_entity_store_with_custom_provider(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
    ) -> None:
        """Test using EntityStore with a custom provider."""
    
>       with ruyi_file.path("ruyipkg_suites", "entities_v0_smoke") as entities_path:

tests/ruyipkg/test_entity_providers.py:54: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4aa12380>, frags = ('ruyipkg_suites', 'entities_v0_smoke'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
___________________________________________________________________________________________________________ test_format_manifest ___________________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4a9f7640>

    def test_format_manifest(ruyi_file: RuyiFileFixtureFactory) -> None:
>       with ruyi_file.path("ruyipkg_suites", "format_manifest") as fixtures_dir:

tests/ruyipkg/test_format_manifest.py:10: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4a9f7640>, frags = ('ruyipkg_suites', 'format_manifest'), @py_assert1 = None, @py_assert4 = None, @py_assert3 = None

    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
        if sys.version_info < (3, 12):
            assert self.module is not None
>       return resources.as_file(resources.files(self.module).joinpath(*frags))
E       TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given

tests/fixtures/__init__.py:21: TypeError
_____________________________________________________________________________________________________________ test_unpack_deb ______________________________________________________________________________________________________________

ruyi_file = <tests.fixtures.RuyiFileFixtureFactory object at 0x7f9b4a9d3100>, ruyi_logger = <ruyi.log.RuyiConsoleLogger object at 0x7f9b4a9d2410>, tmp_path = PosixPath('/tmp/pytest-of-hachi/pytest-19/test_unpack_deb0')

    def test_unpack_deb(
        ruyi_file: RuyiFileFixtureFactory,
        ruyi_logger: RuyiLogger,
        tmp_path: pathlib.Path,
    ) -> None:
>       with ruyi_file.path("cpp-for-host_14-20240120-6_riscv64.deb") as p:

tests/ruyipkg/test_unpack.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.10/contextlib.py:135: in __enter__
    return next(self.gen)
/usr/lib/python3.10/importlib/_common.py:92: in _tempfile
    os.write(fd, reader())
/usr/lib/python3.10/importlib/abc.py:371: in read_bytes
    with self.open('rb') as strm:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <importlib._adapters.DegenerateFiles.Path object at 0x7f9b4a9d2530>, mode = 'rb', args = (), kwargs = {}

    def open(self, mode='rb', *args, **kwargs):
>       raise ValueError()
E       ValueError

/usr/lib/python3.10/importlib/_adapters.py:54: ValueError
========================================================================================================= short test summary info ==========================================================================================================
FAILED tests/pluginhost/test_api.py::test_api_with_ - ValueError
FAILED tests/ruyipkg/test_entity.py::test_entity_store_discovery - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_entity_store_get_entity - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_entity_store_iter_entities - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_entity_store_get_entity_by_ref - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_entity_validation - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_entity_related_refs - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_get_related_entities - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_traverse_related_entities_direct - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_traverse_related_entities_transitive - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_traverse_related_entities_with_type_filter - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_entity_store_is_entity_related_to - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity.py::test_entity_porcelain_output - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_entity_providers.py::test_entity_store_with_custom_provider - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_format_manifest.py::test_format_manifest - TypeError: DegenerateFiles.Path.joinpath() takes 2 positional arguments but 3 were given
FAILED tests/ruyipkg/test_unpack.py::test_unpack_deb - ValueError
====================================================================================================== 16 failed, 23 passed in 0.87s =======================================================================================================
```

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -107,7 +107,13 @@ jobs:
         if: success() && matrix.baseline
         run: |
           pip install -e .
-          pytest
+          if command -v pytest-3 > /dev/null; then
+            # this is the case for Ubuntu python3-pytest
+            pytest-3
+          elif command -v pytest > /dev/null; then
+            # fallback
+            pytest
+          fi
 
       - name: Check for import side effects during CLI startup
         if: success() && !matrix.baseline

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -61,7 +61,7 @@ include = ["ruyi/py.typed"]
 [tool.poetry.group.dev.dependencies]
 mypy = "^1.9.0"
 pyright = "^1.1.389"
-pytest = "^8.2.2"
+pytest = ">=6.2.5"
 ruff = "^0.8.1"
 tomlkit-extras = "^0.2.0"
 typing-extensions = ">=3.10.0.2"

--- a/scripts/install-baseline-deps.sh
+++ b/scripts/install-baseline-deps.sh
@@ -22,15 +22,15 @@ main() {
 
         # for installing ourselves
         python3-pip
+
+        # for running the test suite with purely system deps
+        python3-pytest  # 6.2.5
     )
 
     export DEBIAN_FRONTEND=noninteractive
     export DEBCONF_NONINTERACTIVE_SEEN=true
     sudo apt-get update -qqy
     sudo apt-get install -y "${pkglist[@]}"
-
-    # we need a recent pytest for running the tests though
-    pipx install --system-site-packages pytest
 }
 
 main "$@"

--- a/tests/fixtures/__init__.py
+++ b/tests/fixtures/__init__.py
@@ -1,7 +1,8 @@
-from contextlib import AbstractContextManager
+from contextlib import contextmanager
 from importlib import resources
 import pathlib
 import sys
+from typing import Generator
 
 import pytest
 
@@ -11,26 +12,74 @@ from ruyi.utils.global_mode import GlobalModeProvider
 
 class RuyiFileFixtureFactory:
     def __init__(self, module: resources.Package | None = None) -> None:
+        self._fixtures_dir: pathlib.Path | None = None
+
         if sys.version_info < (3, 12):
             assert module is not None
-        self.module = module
+        # Figure out the fixtures path in a compatible way
+        if module is None:
+            # Python 3.12+ fallback - get the directory of this file
+            self._fixtures_dir = pathlib.Path(__file__).parent
+        elif isinstance(module, str):
+            # Import the module and get its path
+            import importlib
 
-    def path(self, *frags: str) -> AbstractContextManager[pathlib.Path]:
-        if sys.version_info < (3, 12):
-            assert self.module is not None
-        return resources.as_file(resources.files(self.module).joinpath(*frags))
-
-    def plugin_suite(self, suite_name: str) -> AbstractContextManager[pathlib.Path]:
-        if sys.version_info < (3, 12):
-            assert self.module is not None
-        path = resources.files(self.module)
-        if sys.version_info >= (3, 11):
-            path = path.joinpath("plugins_suites", suite_name)
+            mod = importlib.import_module(module)
+            if mod.__file__ is not None:
+                self._fixtures_dir = pathlib.Path(mod.__file__).parent
+            else:
+                self._fixtures_dir = pathlib.Path(__file__).parent
         else:
-            path = path.joinpath("plugins_suites")
-            path = path.joinpath(suite_name)
+            self.module = module
 
-        return resources.as_file(path)
+    @contextmanager
+    def path(self, *frags: str) -> Generator[pathlib.Path, None, None]:
+        if self._fixtures_dir is not None:
+            # directly derive the file path for better compatibility
+            result_path = self._fixtures_dir
+            for frag in frags:
+                result_path = result_path / frag
+            yield result_path
+            return
+
+        # fallback to importlib.resources
+        try:
+            path = resources.files(self.module)
+            for frag in frags:
+                path = path.joinpath(frag)
+            with resources.as_file(path) as p:
+                yield p
+                return
+        except (TypeError, FileNotFoundError):
+            pass
+
+        # final fallback - use the directory of this file
+        result_path = pathlib.Path(__file__).parent
+        for frag in frags:
+            result_path = result_path / frag
+        yield result_path
+
+    @contextmanager
+    def plugin_suite(self, suite_name: str) -> Generator[pathlib.Path, None, None]:
+        if self._fixtures_dir is not None:
+            # directly derive the file path for better compatibility
+            result_path = self._fixtures_dir / "plugins_suites" / suite_name
+            yield result_path
+            return
+
+        # fallback to importlib.resources
+        try:
+            path = resources.files(self.module)
+            path = path.joinpath("plugins_suites").joinpath(suite_name)
+            with resources.as_file(path) as p:
+                yield p
+                return
+        except (TypeError, FileNotFoundError):
+            pass
+
+        # final fallback - use the directory of this file
+        result_path = pathlib.Path(__file__).parent / "plugins_suites" / suite_name
+        yield result_path
 
 
 class MockGlobalModeProvider(GlobalModeProvider):

